src/utils/src/SerialDispatchQueue.c

```diff
@@ -0,0 +1,267 @@
+}
+
+STATUS serialDispatchQueueDispatchTask(SERIAL_DISPATCH_QUEUE_HANDLE handle,
+                                       DispatchQueueTaskFunc dispatchQueueTaskFn,
+                                       UINT64 customData,
+                                       DISPATCH_QUEUE_TASK_PRIORITY priority)
+{
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+    PSerialDispatchQueue pSerialDispatchQueue = NULL;
+    BOOL locked = FALSE;
+    PDispatchQueueTask pDispatchQueueTask = NULL;
+
+    CHK(dispatchQueueTaskFn != NULL, STATUS_NULL_ARG);
+    CHK(IS_VALID_SERIAL_DISPATCH_QUEUE_HANDLE_VALUE(handle), STATUS_INVALID_ARG);
+
+    pSerialDispatchQueue = FROM_SERIAL_DISPATCH_QUEUE_HANDLE(handle);
+
+    CHK(!ATOMIC_LOAD_BOOL(&pSerialDispatchQueue->shutdown), retStatus);
+
+    MUTEX_LOCK(pSerialDispatchQueue->lock);
+    locked = TRUE;
+
+    pDispatchQueueTask = (PDispatchQueueTask) MEMALLOC(SIZEOF(DispatchQueueTask));
+    CHK(pDispatchQueueTask != NULL, STATUS_NOT_ENOUGH_MEMORY);
+
+    pDispatchQueueTask->customData = customData;
+    pDispatchQueueTask->dispatchQueueTaskFn = dispatchQueueTaskFn;
+
+    if (priority == DISPATCH_QUEUE_TASK_PRIORITY_HIGH) {
+        CHK_STATUS(stackQueueEnqueue(pSerialDispatchQueue->taskqueuePriHigh, (UINT64) pDispatchQueueTask));
+    } else {
+        CHK_STATUS(stackQueueEnqueue(pSerialDispatchQueue->taskqueuePriDefault, (UINT64) pDispatchQueueTask));
+    }
+    pDispatchQueueTask = NULL;
+
+    MUTEX_UNLOCK(pSerialDispatchQueue->lock);
+    locked = FALSE;
+
+    CVAR_SIGNAL(pSerialDispatchQueue->executorCvar);
+
+CleanUp:
+
+    if (pDispatchQueueTask != NULL) {
+        MEMFREE(pDispatchQueueTask);
+    }
+
+    if (locked) {
+        MUTEX_UNLOCK(pSerialDispatchQueue->lock);
+    }
+
+    LEAVES();
+    return retStatus;
+}
+
+/////////////////////////////////////////////////////////////////////////////////
+// Internal operations
+/////////////////////////////////////////////////////////////////////////////////
+
+STATUS serialDispatchQueueCreateInternal(PSerialDispatchQueue* ppSerialDispatchQueue)
+{
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+    PSerialDispatchQueue pSerialDispatchQueue = NULL;
+
+    CHK(ppSerialDispatchQueue != NULL, STATUS_NULL_ARG);
+
+    CHK(NULL != (pSerialDispatchQueue = (PSerialDispatchQueue) MEMCALLOC(1, SIZEOF(SerialDispatchQueue))),
+        STATUS_NOT_ENOUGH_MEMORY);
+    pSerialDispatchQueue->lock = MUTEX_CREATE(TRUE);
+    CHK(IS_VALID_MUTEX_VALUE(pSerialDispatchQueue->lock), STATUS_INTERNAL_ERROR);
+    pSerialDispatchQueue->executorCvar = CVAR_CREATE();
+    CHK(IS_VALID_CVAR_VALUE(pSerialDispatchQueue->executorCvar), STATUS_INTERNAL_ERROR);
+    pSerialDispatchQueue->shutdown = FALSE;
+    CHK_STATUS(stackQueueCreate(&pSerialDispatchQueue->taskqueuePriDefault));
+    CHK_STATUS(stackQueueCreate(&pSerialDispatchQueue->taskqueuePriHigh));
+    THREAD_CREATE(&pSerialDispatchQueue->executorTid, serialDispatchQueueExecutor, (PVOID) pSerialDispatchQueue);
```

Comment:
THREAD_DETACH it too

Reply:
I dont need to detach it as the dispatch queue own the executor thread and will join it when dispatch queu is freed.

Reply:
I dont need to detach it. SerialThreadDispatchQueue will join it when the structure get freed.